New guide: AWS DynamoDB with Database Access (#19956)
gravitational/teleport: docs/pages/database-access/guides/aws-dynamodb.mdx

---
title: Database Access with AWS DynamoDB
description: How to access AWS DynamoDB with Teleport Database Access
---

Access to AWS DynamoDB can be provided by [Teleport Database

> Access](../introduction.mdx) for the AWS Console and API. This allows for

Reviewer: ```suggestion
Access](../introduction.mdx). This allows for
```
Same comment as below: This does not provide access to AWS Console/API in general, only DynamoDB specifically.

fine-grain access control through [Teleport's RBAC](../../application-access/introduction.mdx).

This guide will help you to:

- Install the Teleport Database Service.

> - Set up the Teleport Database Service to access the AWS Console and API.

Author: ```suggestion
- Set up the Teleport Database Service to access DynamoDB.
```

If you have not yet deployed the Auth Service and Proxy Service, you should follow
one of our [getting started guides](../getting-started.mdx) or try our Teleport
Application Access [interactive learning track](https://play.instruqt.com/teleport/invite/rgvuva4gzkon).
</Admonition>

We will assume your Teleport cluster is accessible at `teleport.example.com` and
`*.teleport.example.com`. You can substitute the address of your Teleport Proxy
Service. (For Teleport Cloud customers, this will be similar to `mytenant.teleport.sh`.)

<Admonition type="note" title="Example Access Control">
This guide provides an example configuration of IAM access roles as a model,
and uses an EC2 instance to serve the Teleport Database Access service.
The level of access provided may not suit your needs, or may not fit your
organization's access conventions. Adjust as needed to fit your needs.
</Admonition>

## Step 1/4. Create IAM roles for DynamoDB access

The setup described in this guide requires two IAM roles:

- one to identify the EC2 instance running the Teleport Database Access service,
- one applied to users that allows the former role to assume it, and allows

> access to DynamoDB.

Reviewer: I'll leave the structure of this up to you since you prefer the bullets.

I think we should drop the "applied to users" for sure though, and rephrase it similar to "one role that has access to DynamoDB services in AWS".

option, which creates a default trust policy to allow other entities in this
account to assume this role:

![Create Role Step 1](../../../img/database-access/guides/dynamodb-create-role-1.png)

Press "Next". Find the AWS-managed policy `AmazonDynamoDBFullAccess` and then select the policy:

![Create Role Step 2](../../../img/database-access/guides/dynamodb-create-role-2.png)

<Admonition type="note" title="Apply least-privilege permissions">
`AmazonDynamoDBFullAccess` may provide too much access for your intentions. To
use a different IAM policy to reduce permissions, see [Managing access
permissions to your Amazon DynamoDB
Resources](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/access-control-overview.html)
for more details.
</Admonition>

Press "Next". On the next page, enter a role name. In this guide we'll use
the example name `ExampleTeleportDynamoDBRole` for this role.

Under "Select trusted entities", update the JSON to allow the `TeleportDatabaseService`
role to assume this role:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [

>                     "arn:aws:iam::(=aws.aws_access_key=):role/ExampleTeleportDynamoDBRole"

Author: Oof.

                ]
            },
            "Action": "sts:AssumeRole",
            "Condition": {}
        }
    ]
}
```

Finally, click "Create role".

## Step 2/4. Configure the Teleport IAM role mapping

The next step is to give your Teleport users permissions to assume IAM roles in
your Teleport cluster.

You can do this by creating a Teleport role with the `aws_role_arns` field
listing the IAM role ARN created in the previous step. Create a file called
`aws-dynamodb-access.yaml` with the following content:

```yaml
kind: role
version: v5
metadata:
  name: aws-dynamodb-access
spec:
  allow:
    app_labels:
      '*': '*'
    aws_role_arns:
    - arn:aws:iam::(=aws.aws_access_key=):role/ExampleTeleportDynamoDBRole
```

<Details title="Templating aws_role_arns" opened={false}>
The `aws_role_arns` field supports template variables so they can be populated
dynamically based on your users' identity provider attributes. See [Role
Templates](../../access-controls/guides/role-templates.mdx) for details.
</Details>

Create the new role:

```code
$ tctl create -f aws-dynamodb-access.yaml
```

> (!docs/pages/includes/add-role-to-user.mdx role="aws-dynamodb-access"!)

Reviewer: In other database access guides we typically have:
* Step 1: Install Teleport...
* Step 2: Create a Teleport User...
* Step 3: Create a database service configuration ...

and we just instruct them to grant the example user `db_users: '*'` in Teleport RBAC.

Additionally, this RBAC setup example is for application access. In Database access, you only need to allow the user to use a `db_user` that matches the name of the AWS IAM role for DynamoDB - in this example guide that would be `ExampleTeleportDynamoDBRole`. If you use the typical example where we grant `db_users: "*"` in Teleport RBAC, that will allow them to connect as any role for DynamoDB. I suggest sticking to that as it's likely they will choose a different AWS IAM role than the one in our examples.

I would suggest restructuring the guide like this:

```
## Step 1/X. Install Teleport

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

## Step 2/X. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/X. Install the Teleport Database Service
...

## Step 4/X. Create IAM roles for DynamoDB access
...
```

Author: I'm 👍 on restructuring as needed and updating the instructions for database access instead of application access norms. The one thing I'd push back on is the "Create a Teleport user" section. If the only reasoning is 'because we typically do it', I would push back on that. I think it's fair to assume that at this point in the reader's Teleport journey, they already have a Teleport user.  The prerequisites of this guide lay out the assumption that there is already a Teleport cluster in place, implying a user.

Reviewer: Based on our slack discussion (and the issue/PR you are making as a a result of that), we can omit the additional setup steps. Just need to make sure we fix the RBAC steps to be db-access instead of app-access steps in this step now.

Author: I've updated the RBAC example as discussed. Additionally, I've expanded the Prerequisites section of this page to explicitly call for an existing Teleport cluster, with appropriate links. Please take a look, and mark this as resolved if you deem it so.

## Step 3/4. Install the Teleport Database Service

> Create and EC2 instance to host the Teleport Database Service, and attach the
> `TeleportDatabaseService` to it. If you're hosting the service another way, you
> must provide AWS authentication to the host manually.

Reviewer: ```suggestion
Create an EC2 instance to host the Teleport Database Service, and attach the
`TeleportDatabaseService` AWS IAM role to it. If you're hosting the service another way, you
must provide AWS credentials to the service - see [AWS credentials configuration](https://docs.aws.amazon.com/sdkref/latest/guide/creds-config-files.html) for more details.
```

<Admonition type="warning" title="non-standard AWS regions">
For non-standard AWS regions such as AWS GovCloud (US) regions and AWS China
regions, please set the corresponding region in the `AWS_REGION` environment
variable or in the AWS credentials file so that the Application Service can use
the correct STS endpoint.
</Admonition>

### Generate a token

A join token is required to authorize a Teleport Database Service instance
to join the cluster. Generate a short-lived join token and save the output of
the command:

```code

> $ tctl tokens add \
>     --type=db \
>     --db-name=aws-dynamodb

Reviewer: ```suggestion
$ tctl tokens add \
    --type=db \
    --db-name=example-dynamodb
    --db-protocol=dynamodb
```

```

Use the token provided by the output of this command in the next step.

### Install and start Teleport

Install Teleport on the host where you will run the Teleport Application
Service. See our [Installation](../../installation.mdx) page for options
besides Linux servers.

(!docs/pages/includes/install-linux.mdx!)

Edit `/etc/teleport.yaml` to configure the database entry:

```yaml
version: v3
teleport:
  nodename: CHANGEME
  data_dir: /var/lib/teleport
  proxy_server: teleport.example.com
  auth_token: (=presets.tokens.first=)
  ca_pin:
  - (=presets.ca_pin=)
db_service:
  enabled: "yes"
  # Matchers for database resources created with "tctl create" command.
  # For more information: https://goteleport.com/docs/database-access/guides/dynamic-registration/
  resources:
  - labels:
      "*": "*"
  # Lists statically registered databases proxied by this agent.
  databases:

>     - name: "aws-dynamodb"

Reviewer: ```suggestion
    - name: "example-dynamodb"
```
nit: Change this to a name that matches example commands suggested below. I prefer "example-" over "aws-" because it is more clearly an example Teleport DB name.

      protocol: "dynamodb"
      # optional uri, if uri is set then AWS region can be extracted from it
      # or if AWS region is already set then the regions must match.
      # uri: "dynamodb.us-east-1.amazonaws.com:443"
      static_labels:
        env: "dev"
      aws:
        region: "us-east-1"
        account_id: "(=aws.aws_access_key=)"
```

Now start the Teleport Database Service:

<Tabs>
<TabItem label="Package Manager">

```code
$ sudo systemctl start teleport
```

</TabItem>
<TabItem label="TAR archive">

```code
$ sudo teleport install systemd --output=/etc/systemd/system/teleport.service;
$ sudo systemctl enable teleport
$ sudo systemctl start teleport
```

</TabItem>
</Tabs>

## Step 4/4 Connect

Once the Database Service has started and joined the cluster, you can start
connecting to your DynamoDB database.

Create a proxy tunnel:

```code

> $ tsh proxy db --tunnel --port 8000 --db-user=ExampleTeleportDynamoDBRole dynamodb-name

Reviewer: ```suggestion
$ tsh proxy db --tunnel --port 8000 --db-user=ExampleTeleportDynamoDBRole example-dynamodb
```
consistent with example config suggestion above

```

> You can test the connection to the database through the `aws` cli:

Reviewer: You can also use NoSQL Workbench as described in my PR.

Please update "connect-your-client/gui-clients.mdx" to include a section for DynamoDB with NoSQL Workbench.

Reviewer: I suggest including my example of using boto3 python here as well. They can use SDK through the proxy tunnel.

That way we cover:
1. aws cli access
2. GUI client access
3. programmatic access

Author: I've updated `gui-clients.mdx` and linked to it, and provided your example for programmatic access. Please take a look, and mark this as resolved if you deem it so.

---
title: Database Access with AWS DynamoDB
description: How to access AWS DynamoDB with Teleport Database Access
---

Access to AWS DynamoDB can be provided by [Teleport Database
Access](../introduction.mdx). This allows for
fine-grain access control through [Teleport's RBAC](../../database-access/rbac.mdx).

This guide will help you to:

- Install the Teleport Database Service.
- Set up the Teleport Database Service to access DynamoDB.
- Connect to your DynamoDB databases through the Teleport Database Service.

<ScopedBlock scope={["oss", "enterprise"]}>
![DynamoDB Self-Hosted](../../../img/database-access/guides/aws-dynamodb_selfhosted.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>
![DynamoDB Cloud](../../../img/database-access/guides/aws-dynamodb_cloud.png)
</ScopedBlock>

## Prerequisites

- A running Teleport cluster. If you don't already have Teleport running, see
  [How to Choose a Teleport Edition](../../choose-an-edition/introduction.mdx)
  to learn more about your options, and how to set up Teleport.
- AWS account with DynamoDB databases.
- IAM permissions to create IAM roles.

> - `aws` Command Line Interface (CLI) tool installed in PATH.

Reviewer 2: ```suggestion
- `aws` Command Line Interface (CLI) tool installed in `$PATH`.
```

Do you think this makes it more obvious that we're referring to an environment variable? Feel free to ignore this comment if you don't think so.

Author: I have no strong opinion either way, so happy to oblige.

- A host, e.g., an EC2 instance, where you will run the Teleport Database Service.
  This guide assumes an EC2 instance when creating and applying IAM roles, and
  must be adjusted accordingly for custom configurations.

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

<Admonition type="tip" title="Not yet a Teleport user?">
If you have not yet deployed the Auth Service and Proxy Service, you should follow
one of our [getting started guides](../getting-started.mdx) or try our Teleport

> Application Access [interactive learning track](https://play.instruqt.com/teleport/invite/rgvuva4gzkon).

Reviewer: Should be a link to an interactive learning track for Database Access if we have one, otherwise just delete reference to application access learning track

Author: Updated to the correct link.

</Admonition>

We will assume your Teleport cluster is accessible at `teleport.example.com` and

> `*.teleport.example.com`. You can substitute the address of your Teleport Proxy

Reviewer 2: Can we remove the `*.` snippet here? I think that's only relevant for app access.

> We will assume your Teleport cluster is accessible at `teleport.example.com` and
> `*.teleport.example.com`. You can substitute the address of your Teleport Proxy
> Service. (For Teleport Cloud customers, this will be similar to `mytenant.teleport.sh`.)

Author: ```suggestion
```

> We will assume your Teleport cluster is accessible at `teleport.example.com` and
> `*.teleport.example.com`. You can substitute the address of your Teleport Proxy
> Service. (For Teleport Cloud customers, this will be similar to `mytenant.teleport.sh`.)

Author: ```suggestion
We will assume your Teleport cluster is accessible at `teleport.example.com`.
You can substitute the address of your Teleport Proxy Service. (For Teleport Cloud
customers, this will be similar to `mytenant.teleport.sh`.)
```

account to assume this role:

![Create Role Step 1](../../../img/database-access/guides/dynamodb-create-role-1.png)

Click **Next**. Find the AWS-managed policy `AmazonDynamoDBFullAccess` and then select the policy:

![Create Role Step 2](../../../img/database-access/guides/dynamodb-create-role-2.png)

<Admonition type="note" title="Apply least-privilege permissions">
The `AmazonDynamoDBFullAccess` policy may grant more permissions than desired.
If you want to use a different IAM policy to reduce permissions, refer to
[Managing access permissions to your Amazon DynamoDB 
Resources](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/access-control-overview.html)
for more information.
</Admonition>

Click **Next**. On the next page, enter a role name. In this guide we'll use
the example name `ExampleTeleportDynamoDBRole` for this role.

Under "Select trusted entities", update the JSON to allow the `TeleportDatabaseService`
role to assume this role:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [

>                     "arn:aws:iam::(=aws.aws_access_key=):role/TeleportDatabaseService"

Author: We don't have a variable for AWS account id. This renders as `abcd1234-this-is-an-example` so I have to ask, does it really matter what the variable is named in the source?

>                     "arn:aws:iam::(=aws.aws_access_key=):role/TeleportDatabaseService"

Reviewer: could we add an AWS account ID? just make it 12 digits like 123456789012

>                     "arn:aws:iam::(=aws.aws_access_key=):role/TeleportDatabaseService"

Author: Discussed over Slack, it's been agreed.

                ]
            },
            "Action": "sts:AssumeRole",
            "Condition": {}
        }
    ]
}
```

Finally, click **Create Role**.

## Step 2/4. Configure the Teleport IAM role mapping

The next step is to give your Teleport users permissions to assume IAM roles in
your Teleport cluster.

> You can do this by creating a Teleport role with the `aws_role_arns` field

Reviewer: ```suggestion
You can do this by creating a Teleport role with the `db_users` field
```

listing the IAM role ARN created in the previous step. Create a file called
`aws-dynamodb-access.yaml` with the following content:

```yaml
kind: role

> version: v5

Reviewer 2: If this feature is launching with Teleport 12, lets change this to v6. (The default role version will be 6 starting in 12).

If it's going to be backported to 11 and earlier, leaving it at 5 is fine.

Reviewer: Not being backported

> version: v5

Author: ```suggestion
version: v6
```

metadata:
  name: aws-dynamodb-access
spec:
  allow:
    db_labels:
      '*': '*'
    db_users:

>     - '*'

Reviewer: ```suggestion
    - 'ExampleTeleportDynamoDBRole' # only the role name, do not specify the full AWS ARN
```

I realize that I suggested using the wildcard here at some point, but I changed my mind 😅 . I think we should just use the example role we've created in the guide, otherwise there's really no point in telling them to create a role instead of just using the default `access` role (which would already grant these permissions). It also gives us an opportunity to illustrate that they should not use the full role ARN, just the role's name.

```

Create the new role:

```code
$ tctl create -f aws-dynamodb-access.yaml
```

(!docs/pages/includes/add-role-to-user.mdx role="aws-dynamodb-access"!)

## Step 3/4. Install the Teleport Database Service

Create an EC2 instance to host the Teleport Database Service, and attach the
`TeleportDatabaseService` AWS IAM role to it. If you're hosting the service another
way, you must provide AWS credentials to the service - see [AWS credentials
configuration](https://docs.aws.amazon.com/sdkref/latest/guide/creds-config-files.html)
for more details.

<Admonition type="warning" title="non-standard AWS regions">
For non-standard AWS regions such as AWS GovCloud (US) regions and AWS China
regions, please set the corresponding region in the `AWS_REGION` environment

> variable or in the AWS credentials file so that the Application Service can use

Reviewer: ```suggestion
variable or in the AWS credentials file so that the Database Service can use
```

the correct STS endpoint.
</Admonition>

### Generate a token

> A join token is required to authorize a Teleport Database Service instance

Reviewer 2: Since this guide uses an EC2 instance, it's pretty safe to assume the reader is on AWS. IAM joining and EC2 joining are both more secure than static tokens.

I'm fine if we want to keep the static token here since it's less to explain that setting up IAM joining (which has its own guide), but maybe we should just mention there are other more secure methods and soften the "required" wording a bit?

Reviewer: +1 for mentioning the more secure methods, but keeping the static token method for simplicity

Author: Please confirm, are you referring to these two methods?
- https://goteleport.com/docs/server-access/guides/ec2-discovery/
- https://goteleport.com/docs/management/guides/joining-nodes-aws-iam/

Reviewer 2: Yep

Author: See 0fdbfce9c6a73ad77d11b7701b36d4c7ffc73f7d

to join the cluster. Generate a short-lived join token and save the output of
the command:

```code
$ tctl tokens add \
    --type=db \

>     --db-name=example-dynamodb

Reviewer: ```suggestion
    --db-name=example-dynamodb \
```

    --db-protocol=dynamodb
```

Use the token provided by the output of this command in the next step.

### Install and start Teleport

> Install Teleport on the host where you will run the Teleport Application
> Service. See our [Installation](../../installation.mdx) page for options

Reviewer: ```suggestion
Install Teleport on the host where you will run the Teleport Database
Service. See our [Installation](../../installation.mdx) page for options
```

besides Linux servers.

(!docs/pages/includes/install-linux.mdx!)

Edit `/etc/teleport.yaml` to configure the database entry:

```yaml
version: v3
teleport:
  nodename: CHANGEME
  data_dir: /var/lib/teleport
  proxy_server: teleport.example.com
  auth_token: (=presets.tokens.first=)
  ca_pin:
  - (=presets.ca_pin=)
db_service:
  enabled: "yes"
  # Matchers for database resources created with "tctl create" command.
  # For more information: https://goteleport.com/docs/database-access/guides/dynamic-registration/
  resources:
  - labels:
      "*": "*"
  # Lists statically registered databases proxied by this agent.
  databases:
    - name: "example-dynamodb"
      protocol: "dynamodb"
      # optional uri, if uri is set then AWS region can be extracted from it
      # or if AWS region is already set then the regions must match.
      # uri: "dynamodb.us-east-1.amazonaws.com:443"
      static_labels:
        env: "dev"
      aws:
        region: "us-east-1"

>         account_id: "(=aws.aws_access_key=)"

Reviewer 3: same here. a example of  AWS AccountID value should be used here instead of aws access key.

Author: [Same response](https://github.com/gravitational/teleport/pull/19956/files#r1065985779).

---
title: Database Access with AWS DynamoDB
description: How to access AWS DynamoDB with Teleport Database Access
---

Access to AWS DynamoDB can be provided by [Teleport Database
Access](../introduction.mdx). This allows for

> fine-grain access control through [Teleport's RBAC](../../database-access/rbac.mdx).

Author: I'm really not sure about this one. @GavinFrazar I'd welcome your input on this question.

> fine-grain access control through [Teleport's RBAC](../../database-access/rbac.mdx).

Reviewer: I have no preference whether we link between the guides or not. advantages and disadvantages discussion is here:
https://github.com/gravitational/teleport/pull/19956/files#r1093692720

This guide will help you to:

- Install the Teleport Database Service.
- Set up the Teleport Database Service to access DynamoDB.
- Connect to your DynamoDB databases through the Teleport Database Service.

<ScopedBlock scope={["oss", "enterprise"]}>
![DynamoDB Self-Hosted](../../../img/database-access/guides/aws-dynamodb_selfhosted.png)
</ScopedBlock>
<ScopedBlock scope={["cloud"]}>

> ![DynamoDB Cloud](../../../img/database-access/guides/aws-dynamodb_cloud.png)

Reviewer 2: I think we can probably simplify these diagrams and just use a single one. Whether or not the component runs in AWS doesn't really matter, it does the same job either way.

> ![DynamoDB Cloud](../../../img/database-access/guides/aws-dynamodb_cloud.png)

Author: I adjusted the image to have the service in the AWS Cloud. To @zmb3's point it only matters in that the separate images and inclusion of the service in the AWS cloud box help the reader form a mental image of the proposed infrastructure.

</ScopedBlock>

## Prerequisites

> - A running Teleport cluster. If you don't already have Teleport running, see

Author: Adjusted in 70bc2457ccf0a383fa560eba419a3cf5803e4a55

  [How to Choose a Teleport Edition](../../choose-an-edition/introduction.mdx)
  to learn more about your options, and how to set up Teleport.
- AWS account with DynamoDB databases.
- IAM permissions to create IAM roles.
- `aws` Command Line Interface (CLI) tool installed in `$PATH`.
- A host, e.g., an EC2 instance, where you will run the Teleport Database Service.
  This guide assumes an EC2 instance when creating and applying IAM roles, and
  must be adjusted accordingly for custom configurations.

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

> <Admonition type="tip" title="Not yet a Teleport user?">

Author: Removed in 70bc2457ccf0a383fa560eba419a3cf5803e4a55

If you have not yet deployed the Auth Service and Proxy Service, you should follow
one of our [getting started guides](../getting-started.mdx) or try our Teleport
Database Access [interactive learning track](https://play.instruqt.com/teleport/invite/imz8g1n1xzru).
</Admonition>

> We will assume your Teleport cluster is accessible at `teleport.example.com`.

Author: Moved and adjusted in 70bc2457ccf0a383fa560eba419a3cf5803e4a55

You can substitute the address of your Teleport Proxy Service. (For Teleport Cloud
customers, this will be similar to `mytenant.teleport.sh`.)

<Admonition type="note" title="Example Access Control">
This guide provides an example configuration of IAM access roles as a model,
and uses an EC2 instance to serve the Teleport Database Access service.
The level of access provided may not suit your needs, or may not fit your
organization's access conventions. You should adjust the AWS IAM permissions to fit your needs.
</Admonition>

> ## Step 1/4. Create IAM roles for DynamoDB access

Reviewer 2: Doesn't have to be in this PR, but we need to start to do a better job of supporting and documenting how to do things using infra as code tools like Terraform.

Author: @zmb3 I suggest creating a new issue to track this, or linking this comment to an existing issue.

> The setup described in this guide requires two IAM roles: one associated with the EC2 instance running the Teleport Database Access service and another that can be assumed by the EC2 instance role and grants access to DynamoDB services.

Reviewer 4: What does the role assumed by the Database Service do?

Author: Attempted to clarify in cdd7ca0d2c209c2b01d043d9eb46c185926ea818

<Admonition type="note" title="Apply least-privilege permissions">
The `AmazonDynamoDBFullAccess` policy may grant more permissions than desired.
If you want to use a different IAM policy to reduce permissions, refer to
[Managing access permissions to your Amazon DynamoDB 
Resources](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/access-control-overview.html)
for more information.
</Admonition>

Click **Next**. On the next page, enter a role name. In this guide we'll use
the example name `ExampleTeleportDynamoDBRole` for this role.

Under "Select trusted entities", update the JSON to allow the `TeleportDatabaseService`
role to assume this role:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::(=aws.aws_access_key=):role/TeleportDatabaseService"
                ]
            },
            "Action": "sts:AssumeRole",
            "Condition": {}
        }
    ]
}
```

Finally, click **Create Role**.

## Step 2/4. Configure the Teleport IAM role mapping

> The next step is to give your Teleport users permissions to assume IAM roles in

Reviewer 4: Does this sentence mean the same thing as "...give your Teleport users permissions to assume IAM roles when accessing AWS resources registered with Teleport", or something else?

Author: Clarified in 6aa275864c7ce42725be77d3fef8699fd3b90ec0

the correct STS endpoint.
</Admonition>

### Generate a token

<Details title="Alterative methods">

For users with a lot of infrastructure in AWS, or who might create or recreate
many instances, consider alternative methods for joining new EC2 instances running
Teleport:

- [Configure Teleport to Automatically Enroll EC2 instances (Preview)](../../server-access/guides/ec2-discovery.mdx)
- [Joining Nodes via AWS IAM Role](../../management/guides/joining-nodes-aws-iam.mdx)
- [Joining Nodes via AWS EC2 Identity Document](../../management/guides/joining-nodes-aws-ec2.mdx)

</Details>

A join token is required to authorize a Teleport Database Service instance
to join the cluster. Generate a short-lived join token and save the output of
the command:

```code
$ tctl tokens add \
    --type=db \
    --db-name=example-dynamodb \
    --db-protocol=dynamodb
```

Use the token provided by the output of this command in the next step.

### Install and start Teleport

Install Teleport on the host where you will run the Teleport Database
Service. See our [Installation](../../installation.mdx) page for options
besides Linux servers.

(!docs/pages/includes/install-linux.mdx!)

> Edit `/etc/teleport.yaml` to configure the database entry:

Reviewer 4: ```suggestion
Create a file called `/etc/teleport.yaml` with the following content:
```
Since this file doesn't exist yet

```yaml
version: v3
teleport:
  nodename: CHANGEME
  data_dir: /var/lib/teleport

>   proxy_server: teleport.example.com

Reviewer 4: ```suggestion
  proxy_server: teleport.example.com:443
```
Does this default to 3080? In any case, it's probably better to make the port explicit

  auth_token: (=presets.tokens.first=)

>   ca_pin:

Reviewer 2: CA pin isn't really necessary when you connect via the proxy server, as you use the TLS certificate presented by the proxy to establish trust.

Author: Removed.

  - (=presets.ca_pin=)
db_service:
  enabled: "yes"

>   # Matchers for database resources created with "tctl create" command.

Reviewer 2: Are the matchers created with `tctl create` or the database resources? It's not clear the way this is worded.

Author: This language is pulled straight from our example config in `/docs/reference/config`, but I agree that it's inscrutable. I'm going to remove it, and add a link a the end to the referenced page.

  resources:
  - labels:
      "*": "*"
  # Lists statically registered databases proxied by this agent.
  databases:
    - name: "example-dynamodb"
      protocol: "dynamodb"
      # optional uri, if uri is set then AWS region can be extracted from it
      # or if AWS region is already set then the regions must match.
      # uri: "dynamodb.us-east-1.amazonaws.com:443"
      static_labels:
        env: "dev"
      aws:
        region: "us-east-1"
        account_id: "(=aws.aws_access_key=)"
```

Now start the Teleport Database Service:

<Tabs>
<TabItem label="Package Manager">

```code
$ sudo systemctl start teleport
```

</TabItem>
<TabItem label="TAR archive">

```code
$ sudo teleport install systemd --output=/etc/systemd/system/teleport.service;
$ sudo systemctl enable teleport
$ sudo systemctl start teleport
```

</TabItem>
</Tabs>

## Step 4/4 Connect

Once the Database Service has started and joined the cluster, you can start
connecting to your DynamoDB database.

Create a proxy tunnel:

```code
$ tsh proxy db --tunnel --port 8000 --db-user=ExampleTeleportDynamoDBRole example-dynamodb
```

> You can test the connection to the database through the `aws` cli:

Reviewer 4: ```suggestion
You can test the connection to the database through the `aws` CLI:
```

```code
$ aws dynamodb list-tables --endpoint-url=http://localhost:8000
{
    "TableNames": [
        "table1",
        "table2",
        "table3"
    ]
}
```

> You can also connect to this database from the AWS NoSQL Workbench, as documented in our [Database Access GUI Clients](../../connect-your-client/gui-clients.mdx#nosql-workbench) guide.

Author: I'm not opposed to that idea in principle, except that it breaks the convention of having the dedicated page for GUI clients. I could turn that example into a partial to include in both places, but should we instead discuss either doing the same for all of the GUI clients at once, and/or dissolving the separate GUI client page?

    ]
}
```

You can also connect to this database from the AWS NoSQL Workbench, as documented in our [Database Access GUI Clients](../../connect-your-client/gui-clients.mdx#nosql-workbench) guide.

You can also use this tunnel for programmatic access. The example below uses the `boto3` SDK from AWS:

> ```python

Reviewer: ```suggestion
```code
```
or `shell`. Not sure what the docs preference is here, but the snippet example is a shell session

Author: You are technically correct (the best kind of correct), but using `code` results in no syntax highlighting for the python code:

![image](https://user-images.githubusercontent.com/2349184/211604325-b7d3e4c5-fb91-4554-9aea-c65f88ca9dc8.png)

Compared to using `python`:

![image](https://user-images.githubusercontent.com/2349184/211604435-488ebc80-cd9a-475a-8776-86889b653f73.png)

IMO, the highlighting is more useful than having an extra copy button that literally copies `python3` for the reader.
